torch/_functorch/_aot_autograd/collect_metadata_analysis.py

```diff
@@ -91,16 +99,62 @@ def coerce_tangent(x):
         out, "__coerce_tangent_metadata__"
     ):
         out = out.__coerce_tangent_metadata__()  # type: ignore[attr-defined]
-    # It's possible to have a subclass that advertises as contiguous,
-    # but has noncontiguous inner tensors.
-    # Force these to be conntiguous too
+
+    # Coerce subclasses elements to memory_format too.
     if is_traceable_wrapper_subclass(out):
-        for attr in out.__tensor_flatten__()[0]:  # type: ignore[attr-defined]
+        attrs = out.__tensor_flatten__()[0]
+
+        assert isinstance(memory_format, list)
+        for i, attr in enumerate(attrs):
+            elem = getattr(out, attr)
+            new_elem, updated = coerce_tangent(elem, memory_format[1 + i])
+            if updated:
+                setattr(out, attr, new_elem)
+
+    return out, updated
+
+
+# Coercing and collecting traced tangents memory format in one recursive traversal
+# mypy: ignore-errors
+def coerce_tangent_and_suggest_memory_format(
+    x: Tensor, force_memory_format: Optional[torch.memory_format] = None
+):
+    updated = False
+    if not isinstance(x, Tensor):
+        return x, None, updated
+
+    out = x.detach()
+
+    suggest_memory_format = torch._prims_common.suggest_memory_format
+    is_subclass = is_traceable_wrapper_subclass(out)
+
+    memory_format = force_memory_format or suggest_memory_format(out)
+    if not out.is_contiguous(memory_format=memory_format):
+        was = out
+        out = out.contiguous(memory_format=memory_format)
+        updated = out is not was
```

Comment:
nit: if you are checking `out is not was` here, then you don't need the `is_contiguous()` check above: you can just unconditionally call `contiguous()`, and see if it returned the input directly

Reply:
Yeah, was not sure about guarantees of matching is_contiguous/contiguous. But if they match and .contiguous() guaranteed to  return unchanged if is_contiguous() - then we can do it.